File: LeetCode/1-100/Problem4_MedianofTwoSortedArrays/MedianofTwoSortedArrays.py
class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        x, y = len(nums1), len(nums2)
        low, high = 0, x

        while low <= high:
            cutX = (low + high) // 2
            cutY = (x + y + 1) // 2 - cutX

            leftX = float('-inf') if cutX == 0 else nums1[cutX - 1]
            rightX = float('inf') if cutX == x else nums1[cutX]
            leftY = float('-inf') if cutY == 0 else nums2[cutY - 1]
            rightY = float('inf') if cutY == y else nums2[cutY]

            if leftX <= rightY and leftY <= rightX:
                if (x + y) % 2 == 0:
                    return (max(leftX, leftY) + min(rightX, rightY)) / 2.0
                else:
                    return max(leftX, leftY)
            elif leftX > rightY:
                high = cutX - 1
            else:
                low = cutX + 1

File: LeetCode/1-100/Problem4_MedianofTwoSortedArrays/MedianofTwoSortedArrays.go
package main

import (
	"math"
)

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	// Ensure nums1 is the smaller array
	if len(nums1) > len(nums2) {
		return findMedianSortedArrays(nums2, nums1)
	}

	x, y := len(nums1), len(nums2)
	low, high := 0, x

	for low <= high {
		partitionX := (low + high) / 2
		partitionY := (x+y+1)/2 - partitionX

		maxLeftX := math.Inf(-1)
		if partitionX != 0 {
			maxLeftX = float64(nums1[partitionX-1])
		}

		minRightX := math.Inf(1)
		if partitionX != x {
			minRightX = float64(nums1[partitionX])
		}

		maxLeftY := math.Inf(-1)
		if partitionY != 0 {
			maxLeftY = float64(nums2[partitionY-1])
		}

		minRightY := math.Inf(1)
		if partitionY != y {
			minRightY = float64(nums2[partitionY])
		}

		if maxLeftX <= minRightY && maxLeftY <= minRightX {
			if (x+y)%2 == 0 {
				return (math.Max(maxLeftX, maxLeftY) + math.Min(minRightX, minRightY)) / 2.0
			} else {
				return math.Max(maxLeftX, maxLeftY)
			}
		} else if maxLeftX > minRightY {
			high = partitionX - 1
		} else {
			low = partitionX + 1
		}
	}

}
